### ISSUE TITLE
functions.yaml is incomplete

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1828]
Not all public functions are listed in `functions.yaml`. This issue is to:

 - Identify what is missing.
 - Add all missing functions to `functions.yaml`

**Comment by szhorvat** [comment]
This is a rough first list, first auto-generated, then lightly pruned down manually. Some more entries likely need to be removed.

- [ ] igraph_automorphism_group
- [ ] igraph_average_local_efficiency
- [ ] igraph_average_path_length_dijkstra
- [ ] igraph_bfgs
- [ ] igraph_bfs_simple
- [ ] igraph_bipartite_game
- [ ] igraph_bridges
- [ ] igraph_clique_size_hist
- [ ] igraph_delete_vertices_idx
- [ ] igraph_deterministic_optimal_imitation

- [ ] igraph_eigen_laplacian
- [ ] igraph_eigen_matrix
- [ ] igraph_eigen_matrix_symmetric
- [ ] igraph_erdos_renyi_game
- [ ] igraph_even_tarjan_reduction
- [ ] igraph_feedback_arc_set
- [ ] igraph_from_prufer
- [ ] igraph_get_adjacency_sparse
- [ ] igraph_get_eids_multi
- [ ] igraph_get_shortest_path
- [ ] igraph_get_shortest_path_bellman_ford
- [ ] igraph_get_shortest_path_dijkstra
- [ ] igraph_get_sparsemat
- [ ] igraph_global_efficiency
- [ ] igraph_gomory_hu_tree
- [ ] igraph_harmonic_centrality
- [ ] igraph_harmonic_centrality_cutoff
- [ ] igraph_has_loop
- [ ] igraph_hrg_resize
- [ ] igraph_hrg_sample
- [ ] igraph_hrg_size
- [ ] igraph_hub_and_authority_scores
- [ ] igraph_induced_subgraph_map
- [ ] igraph_is_bigraphical
- [ ] igraph_is_graphical
- [ ] igraph_is_same_graph
- [ ] igraph_is_tree
- [ ] igraph_isomorphic_function_vf2
- [ ] igraph_largest_weighted_cliques
- [ ] igraph_lcf
- [ ] igraph_local_efficiency
- [ ] igraph_maximal_cliques_hist
- [ ] igraph_maximal_cliques_subset
- [ ] igraph_maximum_matching
- [ ] igraph_moran_process
- [ ] igraph_personalized_pagerank_vs

- [ ] igraph_pseudo_diameter_dijkstra
- [ ] igraph_random_edge_walk

- [ ] igraph_random_spanning_tree
- [ ] igraph_realize_degree_sequence
- [ ] igraph_residual_graph (do we need this at all?)
- [ ] igraph_reverse_residual_graph (do we need this at all?)
- [ ] igraph_rewire_directed_edges
- [ ] igraph_roulette_wheel_imitation
- [ ] igraph_similarity_dice_es
- [ ] igraph_similarity_dice_pairs
- [ ] igraph_similarity_jaccard_es
- [ ] igraph_similarity_jaccard_pairs
- [ ] igraph_simplify_and_colorize


- [ ] igraph_st_mincut

- [ ] igraph_stochastic_imitation
- [ ] igraph_subisomorphic
- [ ] igraph_subisomorphic_function_vf2
- [ ] igraph_to_prufer
- [ ] igraph_transitive_closure_dag
- [ ] igraph_tree_game
- [ ] igraph_trussness
- [ ] igraph_version
- [ ] igraph_vertex_coloring_greedy
- [ ] igraph_weighted_clique_number
- [ ] igraph_weighted_cliques
- [ ] igraph_weighted_sparsemat
- [ ] igraph_zeroin

**Comment by szhorvat** [comment]
The above was generated as follows. First, we get the public symbols:

```
nm -gU src/libigraph.dylib > syms.txt
```

We convert to YAML to JSON:

```
yq eval -o=json functions.yaml
```

Then in Mathematica:

```
interfaceFunctions = 
  Import["~/Repos/igraph-main/igraph/interfaces/functions.json", 
   "RawJSON"];

publicFunctions = Cases[
     Import["~/Repos/igraph-main/igraph/build/syms.txt", "Table", 
       "Numeric" -> False][[All, {2, 3}]],
     {"T", _}
     ][[All, 2]] // Map[StringDrop[#, 1] &];

missingFunctions = DeleteCases[
   Complement[publicFunctions, Keys[interfaceFunctions]],
   s_ /; 
    StringContainsQ[s, 
     "igraph_2wheap" | "igraph_array3" | "igraph_vector" | 
      "igraph_matrix" | "igraph_heap" | "igraph_stack" | 
      "igraph_dqueue" | "igraph_d_indheap" | "igraph_hashtable" | 
      "igraph_set_" | "igraph_estack_" | "cattribute" | 
      "igraph_complex" | "igraph_attribute_combination" | 
      "igraph_blas" | "igraph_lapack" | "IGRAPH_" | "igraph_i_" | 
      "igraph_marked_queue" | "igraph_psumtree" | "igraph_sparsemat" |
       "igraph_spmatrix" | "igraph_strvector" | "igraph_rng_" | 
      "print" | "igraph_trie" | "igraph_es_" | "igraph_ess_" | 
      "igraph_vs_" | "igraph_vss_" | "igraph_vit_" | "igraph_eit_" | 
      "igraph_error" | "igraph_warning" | "igraph_fatal" | 
      "igraph_adjlist_" | "igraph_inclist_" | "igraph_lazy_adjlist_" |
       "igraph_lazy_inclist_" | "_callback" | "_handler_" | 
      "_destroy" | "_init"]
   ];

StringJoin[ "- [ ] " <> # <> "\n" & /@ LexicographicSort[missingFunctions]]
```

**Comment by GroteGnoom** [comment]
Should `igraph_arpack_options_get_default()` be exported? Comments on other arpack functions suggest not:
>    \# TODO(ntamas): this should probably not be exposed to higher-level
    # interfaces; igraph's goal is not to provide an ARPACK wrapper

**Comment by ntamas** [comment]
I don't think it should be exported in higher-level interfaces, but I think it's useful to have it in the C core. Most ARPACK functions that require an `igraph_arpack_options_t` should be able to accept NULL there and just use the "default" ARPACK options (I don't know whether this is already implemented, I remember that we planned this). Allowing the user to get a pointer to the default options will let the user tweak the default settings if he is working with a problem that necessitates this.

**Comment by GroteGnoom** [comment]
What do the `*` mean? Like after igraph_circulant?

**Comment by szhorvat** [comment]
It means priority: it makes sense to add these first. This is based on my own judgement. Most of these are newly added functions.

**Comment by ntamas** [comment]
Also, note that not all functions need to be added. I'd say only those functions need to be placed in `functions.yaml` where there is a reason to generate glue code for a high-level interface. E.g., `igraph_almost_equals` is probably not needed in a high-level interface, while `igraph_circulant` clearly is.

**Comment by GroteGnoom** [comment]
Now we're left with 
- functions with varargs like ` igraph_progressf` and ` igraph_statusf`, which I wasn't sure how to handle
- igraph_zeroin and igraph_bfgs, which I wasn't sure should even be in igraph
- deprecated functions

**Comment by ntamas** [comment]
`igraph_zeroin` and `igraph_bfgs` shouldn't be exposed either.
As for deprecated functions, I think we shouldn't expose them, except for those that are already used by the R interface -- but if they were used already, then they are already in `functions.yaml`, so again, no action is needed now.

**Comment by ntamas** [comment]
@szhorvat Can you re-generate the list please one (hopefully) last time with Mathematica? Just to see if there's something new that we might have missed.

**Comment by szhorvat** [comment]
- [ ] igraph_full_multipartite
- [ ] igraph_get_widest_path
- [ ] igraph_get_widest_paths
- [ ] igraph_graph_center
- [ ] igraph_graph_count
- [ ] igraph_layout_umap
- [ ] igraph_layout_umap_3d
- [ ] igraph_plfit_result_calculate_p_value
- [ ] igraph_read_graph_dimacs
- [ ] igraph_similarity_jaccard_es ??
- [ ] igraph_sparse_adjacency
- [ ] igraph_sparse_weighted_adjacency
- [ ] igraph_turan
- [ ] igraph_vertex_path_from_edge_path
- [ ] igraph_widest_paths_dijkstra
- [ ] igraph_widest_paths_floyd_warshall
- [ ] igraph_write_graph_dimacs
